### PR TITLE
Fix AssertionErrors after purging events

### PR DESCRIPTION
**Pull request by richvdh**
If you purged a bunch of events from your database, and then restarted synapse
without receiving more events, then you would get a bunch of AssertionErrors on
restart.

This fixes the situation by rewinding the stream processors.

(I've also updated the changelog for #11536, since this was the cause of the mixup I was trying to debug there.)

### CODE REVIEW

File: scripts-dev/check-newsfragment [hunk 1]
```diff
@@ -42,8 +42,8 @@ echo "--------------------------"
 echo
 
 matched=0
-for f in $(git diff --name-only FETCH_HEAD... -- changelog.d); do
-    # check that any modified newsfiles on this branch end with a full stop.
+for f in $(git diff --diff-filter=d --name-only FETCH_HEAD... -- changelog.d); do
```
Comment (richvdh) [inline scripts-dev/check-newsfragment:45]: `--diff-filter=d` tells git to ignore deleted files here. This should probably be a separate PR but at this stage I've got about 15 of the things open :-S

File: scripts-dev/check-newsfragment [hunk 2]
```diff
@@ -42,8 +42,8 @@ echo "--------------------------"
...
+    # check that any added newsfiles on this branch end with a full stop.
```
Comment (clokep) [inline scripts-dev/check-newsfragment:46]: So this is added or modified, but not deleted?

File: synapse/handlers/stats.py [hunk 3]
```diff
@@ -80,6 +80,17 @@ async def _unsafe_process(self) -> None:
...
+            if self.pos > room_max_stream_ordering:
+                # apparently, we've processed more events than exist in the database!
+                # this can happen if events are removed with history purge or similar.
+                logger.warning(
+                    "Event stream ordering appears to have gone backwards (%i -> %i): "
+                    "rewinding stats processor",
+                    self.pos,
+                    room_max_stream_ordering,
+                )
+                self.pos = room_max_stream_ordering
```
Comment (clokep) [inline synapse/handlers/stats.py:93]: I'm having trouble convincing myself this is a "safe" operation to do. I thought the max stream ordering was monotonically increasing, even if you purged events but I might be misunderstanding how it is used?

File: scripts-dev/check-newsfragment (hunk 2 above)
Comment (richvdh) [reply scripts-dev/check-newsfragment:46]: yes.

File: synapse/handlers/stats.py (hunk 3 above)
Comment (richvdh) [reply synapse/handlers/stats.py:93]: well, that's a fair question.

The fact is that currently, on sqlite, we *do* reuse stream orderings, because we have no way of knowing which stream orderings were used before synapse was restarted. 

For things like the stats and userdirectory, that's not really a problem, because any events that get a stream ordering later than the max at startup must be new events, so I think this must be safe.

One place it *is* a problem is in client /sync - since stream_orderings are encoded in sync tokens, any events that reuse old stream_orderings probably won't get synced to the client. Ironically, we *don't* log any warnings in that case.

So, sure, a better solution here is not to reuse stream orderings at all, but that's not easy. Meanwhile I think we may as well update these bits to stop them complaining.

File: scripts-dev/check-newsfragment (hunk 2 above)
Comment (richvdh) [reply scripts-dev/check-newsfragment:46]: see https://git-scm.com/docs/git-diff#Documentation/git-diff.txt---diff-filterACDMRTUXB82308203 for docs on this btw.

**Review by clokep — APPROVED** [review]
